identifier/identifier.go

```diff
@@ -0,0 +1,31 @@
+// The identifier package defines types for RFC 8555 ACME identifiers.
+package identifier
+
+// IdentifierType is a named string type for registered ACME identifier types.
+// See https://tools.ietf.org/html/rfc8555#section-9.7.7
+type IdentifierType string
+
+const (
+	// IdentifierDNS is specified in RFC 8555 for DNS identifiers.
+	IdentifierDNS = IdentifierType("dns")
+)
+
+// ACMEIdentifier is a struct encoding an identifier that can be validated. The
+// protocol allows for different types of identifier to be supported (DNS
+// names, IP addresses, etc.), but currently we only support RFC 8555 DNS type
+// identifiers for domain names.
+type ACMEIdentifier struct {
+	// Type is the registered IdentifierType of the identifier.
+	Type IdentifierType `json:"type"`
+	// Value is the value of the identifier. For a DNS type identifier it is
+	// a domain name.
+	Value string `json:"value"`
+}
+
+// TODO(@cpu): DNSIdentifier doc string
+func DNSIdentifier(value string) ACMEIdentifier {
```

Comment:
We have a ton of test code that implements this as a one-off function. It seems worth promoting into an exported `identifier` function. I haven't done a pass yet through the existing code to use `identifier.DNSIdentifier` instead of one-offs.